ml-prims/src/common/device_buffer.hpp

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../utils.h"
+
+#include "cuml_allocator.hpp"
+
+namespace MLCommon {
+
+/**
+ * RAII object owning a contigous typed device buffer. The passed in allocator supports asynchronus allocation and
+ * deallocation so this can be used for temporary memory 
+ * @code{.cpp}
+ * template<typename T>
+ * void foo( const cumlHandle_impl& h, ..., cudaStream_t stream )
+ * {
+ *     ...
+ *     device_buffer<T> temp( h.getDeviceAllocator(), stream, 0 )
+ *     
+ *     temp.resize(n, stream);
+ *     kernelA<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     kernelB<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     temp.release(stream);
+ * }
+ * @endcode
+ */
+template<typename T>
+class device_buffer
```

Comment:
I am really enjoying reading your proposal. It is a cool stuff!
- One minor question: does it make sense to have a base class of `device_buffer` and `host_buffer` to reduce the duplicate code?

Reply:
Thanks for the feedback. That is a good suggestion. I will add a base class for `device_buffer` and `host_buffer` with the next iteration of this.